eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Bbody.hs

```diff
@@ -36,77 +60,265 @@ import Cardano.Ledger.Conway.Rules.Ledgers ()
 import Cardano.Ledger.Conway.Rules.Utxo (ConwayUtxoPredFailure)
 import Cardano.Ledger.Conway.Rules.Utxos (ConwayUtxosPredFailure)
 import Cardano.Ledger.Conway.Rules.Utxow (ConwayUtxowPredFailure)
+import Cardano.Ledger.Conway.UTxO (txNonDistinctRefScriptsSize)
 import Cardano.Ledger.Core
+import qualified Cardano.Ledger.Era as Era
+import Cardano.Ledger.Keys (DSignable, Hash)
+import Cardano.Ledger.Shelley.LedgerState (LedgerState (..), lsUTxOState, utxosUtxo)
 import Cardano.Ledger.Shelley.Rules (
-  ShelleyBbodyPredFailure (..),
-  ShelleyLedgersPredFailure (..),
+  BbodyEnv (..),
+  ShelleyBbodyEvent (..),
+  ShelleyBbodyPredFailure,
+  ShelleyBbodyState (..),
+  ShelleyLedgersEnv (..),
+  ShelleyLedgersPredFailure,
   ShelleyPoolPredFailure,
   ShelleyUtxoPredFailure,
   ShelleyUtxowPredFailure,
  )
+import qualified Cardano.Ledger.Shelley.Rules as Shelley (ShelleyBbodyPredFailure (..))
+import Control.State.Transition (
+  Embed (..),
+  STS (..),
+  TRC (..),
+  TransitionRule,
+  judgmentContext,
+  (?!),
+ )
+import Data.Foldable (Foldable (foldMap'))
+import Data.Monoid (Sum (getSum))
+import qualified Data.Monoid as Monoid (Sum (..))
+import Data.Sequence (Seq)
+import GHC.Generics (Generic)
+import NoThunks.Class (NoThunks (..))
+
+data ConwayBbodyPredFailure era
+  = WrongBlockBodySizeBBODY
+      -- | Actual Body Size
+      !Int
+      -- | Claimed Body Size in Header
+      !Int
+  | InvalidBodyHashBBODY
+      -- | Actual Hash
+      !(Hash (EraCrypto era) EraIndependentBlockBody)
+      -- | Claimed Hash
+      !(Hash (EraCrypto era) EraIndependentBlockBody)
+  | -- | LEDGERS rule subtransition Failures
+    LedgersFailure !(PredicateFailure (EraRule "LEDGERS" era))
+  | TooManyExUnits
+      -- | Computed Sum of ExUnits for all plutus scripts
+      !ExUnits
+      -- | Maximum allowed by protocal parameters
+      !ExUnits
+  | RefScriptsSizeTooBig
```

Comment:
Let's add a prefix `Body`. since we'll have a similar predicate failure on the Tx level
```suggestion
  | BodyRefScriptsSizeTooBig
```

Reply:
Done